### PR TITLE
Avoid query cache sharding code in single-threaded mode

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
In non-parallel compilers, this is just adding needless overhead at compilation time (since there is only one shard statically anyway). This amounts to roughly ~10 seconds reduction in bootstrap time, with overall neutral (some wins, some losses) performance results.

Parallel compiler performance should be largely unaffected by this PR; sharding is kept there.

### CODE REVIEW

File: compiler/rustc_query_system/src/query/plumbing.rs [hunk 1]
```diff
@@ -107,13 +92,11 @@ where
     ) -> Option<()> {
         // We use try_lock_shards here since we are called from the
```
Comment (joshtriplett) [inline compiler/rustc_query_system/src/query/plumbing.rs:93]: This comment no longer matches the code.

**Comment by joshtriplett** [comment]
We absolutely should make changes that improve the performance of the non-parallel compiler. However, it'd be nice if we had some means of measuring the impact on the parallel compiler. Is there any tracking issue for having rustc-perf measure parallel compiler performance?

**Comment by joshtriplett** [comment]
> The expectation is not necessarily to land this code as-is (I think that's unlikely), but to identify how much of a win this is

Ah, got it. In that case, any objections to marking this PR as a draft? That often serves as a good indicator of "this is being used to check performance of an idea".

I absolutely agree that we shouldn't run that tracking in general on every perf run until we have more active development on it. But it'd help to have the ability to enable it, and to be able to run it specifically for PRs we'd expect to affect it. (As well as, perhaps, a perf run per release.)

**Comment by Mark-Simulacrum** [comment]
The lack of an assigned reviewer (i.e., explicit r? @ghost) is my signal for whether work is not intended for review -- I don't typically use the draft view on GitHub, I don't really care either way though.

Tracking it even irregularly still requires quite a bit of work to get all the pieces in the right order today, but it's not something necessarily blocked on infra work (try builds with the right CI changes are sufficient), so someone well-motivated could start doing so.

**Comment by Mark-Simulacrum** [comment]
FWIW, one reason I am reluctant to track is that we *already* cannot really reliably keep up with triaging numerous, typically relatively small, perf regressions. I suspect that the parallel compiler mode will be even more difficult to diagnose regressions in -- at least in the current suite of tools -- so I am reluctant to add that extra data to our perf-tracking work.

**Comment by joshtriplett** [comment]
Ah, sorry, missed the `r? @ghost`.

Good to know that it's something someone *could* put together if motivated to do so.

> FWIW, one reason I am reluctant to track is that we already cannot really reliably keep up with triaging numerous, typically relatively small, perf regressions. I suspect that the parallel compiler mode will be even more difficult to diagnose regressions in -- at least in the current suite of tools -- so I am reluctant to add that extra data to our perf-tracking work.

I absolutely wouldn't expect the perf team to handle diagnosing or dealing with such regressions; the only time we'd want to consider them is when making changes like this that may trade off optimization of one for the other.

**Comment by klensy** [comment]
I've done some work in #93787 about separating parallel_compiler, but it need review.

**Comment by Mark-Simulacrum** [comment]
I think this would be largely orthogonal from that PR (or increase the work needed to separate it out), since it's about moving an API from being largely equivalent across the question of parallel compiler to being quite different.

**Comment by nnethercote** [comment]
> My initial thinking is we can either just land this (pretty much as-is, modulo some further comment cleanup / renaming struct variables, etc.) or try to cfg gate all the sharding away.

The former would almost certainly regress the parallel compiler, right? Given that all the shards would be locked where currently a single shard is locked.

**Comment by Mark-Simulacrum** [comment]
Presuming there's heavy contention on a given query -- yes. It's worth noting that each individual query still has its own lock, so if threads are doing parallel work and largely executing distinct queries, then contention is probably minimal. We don't really hold the locks themselves for that long, either. https://github.com/rust-lang/rust/pull/61779 added the sharding based on what looks like ~1 data point (at least documented), though the 30% win there is certainly significant. On the other hand, IIRC 1st-gen Ryzen was particularly bad at latency when shuffling cache lines between cores, so I'm not sure how much of a win this ends up being.

I think it's pretty likely that we could fairly minimally adjust the PR to have `Sharded<T>` be just `Lock<T>` on non-parallel compilers and `[Lock<T>; N]` on parallel compilers, with functions appropriately mapping to each use case around that. I can try to experiment with that and make a concrete proposal (i.e., delta on this PR), though I admit that a good part of me sort of wants to just not bother with the cfgs across rustc_query crates necessary to make that happen.

**Comment by bjorn3** [comment]
Doesn't non-parallel rustc already use a single shard? https://github.com/rust-lang/rust/blob/3b186511f62b0ce20e72ede0e8e13f8787155f02/compiler/rustc_data_structures/src/sharded.rs#L18-L19

**Comment by Mark-Simulacrum** [comment]
Yes, it does. As the perf results illustrate, the layers of abstraction there do add a fairly considerable chunk of compilation time, though runtime performance is largely unaffected.

I'm working on a revision of this patch that tries to aim to cfg the sharding more carefully alongside some cleanups so we keep equivalent results on parallel builds.

**Comment by Mark-Simulacrum** [comment]
@bors try @rust-timer queue

Alright, pushed up a new set of commits which do a more thorough refactoring split across multiple commits and keep largely identical high-level behavior for parallel compilation (modulo a few mostly minor details around QueryLookup keeping shard indices and such, rather than recomputing them from scratch. Those are hard to cfg pipe around and do not feel likely to be meaningful to me).

**Comment by Mark-Simulacrum** [comment]
Results look pretty mixed but I think overall neutral -- stress tests dominate regressions and there are also some small improvements; looking at cachegrind diffs locally it doesn't look like they're obviously related to the work in this PR, so I am marking the regression as triaged.

**Comment by cjgillot** [comment]
I'm wondering: is there a longer-term context for these changes? This PR optimizes the serial compiler and degrades the parallel compiler. Should we consider dropping/reimplementing the parallel compiler?

**Comment by Mark-Simulacrum** [comment]
The latest version of this PR should have roughly neutral effect on parallel compiler performance, since it keeps sharding things in that mode.

IMO, it may not be a bad idea to drop the parallel compiler support unless we have concrete investment expected in the next 6-18 month timeframe, since it does cause constant 'small' pain across many bits of the compiler. But this PR would ideally not be blocked on a decision there :)

**Comment by lqd** [comment]
> Should we consider dropping/reimplementing the parallel compiler?

The parallel compiler currently suffers from non-deterministic ICEs (in addition to the other known issues about jobserver pipe contention, lack of horizontal scalability, etc) but when/if it works, it seems to be surprisingly effective on compile times.

**Comment by cjgillot** [comment]
LGTM.
Why did you remove the caching of the key hash using the `QueryLookup` type? Could we gain a bit of perf by avoiding to hash keys multiple times?
r=me either way

**Comment by Mark-Simulacrum** [comment]
It was already unused -- if you look at the commit deleting QueryLookup, we're not actually threading it down anywhere. The query shard was used, but recomputing it from the hash we calculate anyway should be pretty cheap (it's just a shift and mask) and threading it through just on parallel compilers seems like more work than we ought to do.

It's possible actually caching the key hash would make sense but I think we see a little benefit from saving some registers/stack allocation to thread the data down too, so it's not guaranteed to help. I expect the query key hash is typically really fast to compute, as the majority of our keys are e.g. DefId or so, which take just a handful of instructions to compute FxHash for.

@bors r=cjgillot

**Pull request merged**